src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Net.Http;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+using Microsoft.AspNetCore.Server.Kestrel.Core.WebTransport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.WebTransport;
+
+/// <summary>
+/// Controls the WebTransport session and keeps track of all the streams.
+/// </summary>
+internal class WebTransportSession : IWebTransportSession
+{
+    private readonly Dictionary<long, WebTransportStream> _openStreams = new();
+    private readonly ConcurrentQueue<WebTransportStream> _pendingStreams = new();
+    private readonly SemaphoreSlim _pendingAcceptStreamRequests = new(0);
+    private readonly Http3Connection _connection;
+    private readonly Http3Stream _controlStream = default!;
```

Comment:
Control Stream means something specific in the Http3 spec that doesn't match what you're doing here. This is the original Connect stream.